inverse/beamformer_sam.m

```diff
@@ -123,136 +156,164 @@
+        ft_error(sprintf('unknown fixedori method %s', fixedori));
     end
-
-    % The optimum orientation is the eigenvector that corresponds to the
-    % smallest eigenvalue.
-
-    % Double check that this is the case, because for single sphere head
-    % model, one of the eigenvectors corresponds to the radial direction,
-    % giving lead fields that are zero (to within machine precission).
-    % The eigenvalue corresponding to this eigenvector can actually be
-    % the smallest and can give the optimum (but wrong) Z-value!)
-
-    ori1 = U(:,1); ori1 = ori1/norm(ori1);
-    ori2 = U(:,2); ori2 = ori2/norm(ori2);
-    % ori3 = U(:,3); ori3 = ori3/norm(ori3);
-
-    L1 = L * ori1;
-    L2 = L * ori2;
-    % L3 = L * ori3;
-
-    if (norm(L1)/norm(L2)) < 1e-6
-      % the first orientation seems to be the silent orientation
-      % use the second orientation instead
-      opt_vox_or = ori2;
-    else
-      opt_vox_or = ori1;
+      
+    % convert the U matrix into the optimal orientation vector
+    switch fixedori
+      case 'spinning'
+        % do nothing, optimum orientation is already computed above
+      otherwise
+        
+        % The optimum orientation is the eigenvector that corresponds to the
+        % biggest eigenvalue (biggest value is more logical, as it relates to SNR).
+        
+        if isfield(headmodel,'singlesphere') && ~strcmp(fixedori,'moiseev')
+          % If baseline noise covariance is not used, for single sphere head
+          % model, one of the eigenvectors corresponds to the radial direction,
+          % giving lead fields that are zero (to within machine precission).
+          % The eigenvalue corresponding to this eigenvector can actually be
+          % the biggest and can give the optimum (but wrong) Z-value!)
+          ori1 = U(:,1); ori1 = ori1/norm(ori1);
+          ori2 = U(:,2); ori2 = ori2/norm(ori2);
+          % ori3 = U(:,3); ori3 = ori3/norm(ori3);
+          
+          L1 = lf * ori1;
+          L2 = lf * ori2;
+          
+          if (norm(L1)/norm(L2)) < 1e-6
+            % the first orientation seems to be the silent orientation
+            % use the second orientation instead
+            opt_vox_or = ori2;
+          else
+            opt_vox_or = ori1;
+          end
+          
+        else
+          % select eigenvector with biggest eigenvalue
+          [~, ori_inx] = sort(diag(S), 'descend');
+          ori = U(:,ori_inx(1));
+          opt_vox_or = ori/norm(ori);
+        end
     end
-
-  end, % if fixedori
-
-  % compute the spatial filter for the optimal source orientation
-  gain        = lf * opt_vox_or;
-  trgain_invC = gain' * inv_cov;
-  SAMweights  = trgain_invC / (trgain_invC * gain);
-
+    dipout.ori{diplop}    = opt_vox_or;
+       
+    % compute the spatial filter for the optimal source orientation
+    gain        = lf * opt_vox_or;
+    trgain_invC = gain' * inv_cov;
+    SAMweights  = trgain_invC / (trgain_invC * gain);  
```

Comment:
if strcmp(fixedori,'moiseev')
        % get pseudoZ
        Ng                      = gain' * Nproj * gain;
        Sg                      = gain' * Sproj * gain; 
        dipout.pseudoZ{diplop}  = Sg / Ng;
    end

Reply:
Hi Adonay, it's strange that you cannot PR on this PR, but I don't mind adding this to the code. Just for my understanding, this is only applicable to the 'moiseev' method? Why then does it need to be handled at this location separately? I understand it given the current formulation, but isn't Ng./Sg equal to Nproj./Sproj. Both Ng, Nproj etc. seem to be scalars, and the 'gain' variable is a vector (thus gain'*gain is a scalar, and that drops out of the division). Would it be OK for you if I compute the pseudoZ higher up in the code, by a direct division between Nproj and Sproj?

Reply:
Nope. Ng and Sg are the noise and power projected by the most salient LF direction, they are scalars. Nproj and Sproj are n_sensor x n_sensors, expressing the noise (by inv cov matrix projection) and the signal (in case of avg BF the average signal by inv cov matrix projection or the signal inv covariance). By matrix projection I mean A*B*A.
And for each LF there will be a different pseudoZ, e.g. rest and finger tapping, the pseudoZ map will show activation in the motor area as a ratio finger/rest pseudoZ.
Yes, it's only applicable to 'moiseev' method, if the noise covariance is diagonal (as in the other methods) then pseudoZ will just show the pseudo magnitude of the source.

Reply:
My bad, of course you're totally right. Sorry for my dimwittedness and thanks for your patience.